File: library/src/main/java/me/shkschneider/skeleton/java/ReflectHelper.java
package me.shkschneider.skeleton.java;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import me.shkschneider.skeleton.helper.LogHelper;

@Deprecated // Discouraged
public class ReflectHelper {

    public static class Field {

        @Nullable
        public static List<String> fields(@NonNull final Object object) {
            try {
                final List<String> fields = new ArrayList<>();
                for (final java.lang.reflect.Field field : object.getClass().getDeclaredFields()) {
                    fields.add(field.getName());
                }
                return fields;
            }
            catch (final Exception e) {
                LogHelper.wtf(e);
            }
            return null;
        }

        @Nullable
        public static Object field(@NonNull final Object object, @NonNull final String declaredField) {
            try {
                java.lang.reflect.Field field = object.getClass().getDeclaredField(declaredField);
                field.setAccessible(true);
                return field.get(object);
            }
            catch (final Exception e) {
                LogHelper.wtf(e);
            }
            return null;
        }

        @Nullable
        public static boolean field(@NonNull final Object object, @NonNull final String declaredField, final Object value) {
            try {
                java.lang.reflect.Field field = object.getClass().getDeclaredField(declaredField);
                field.setAccessible(true);
                field.set(object, value);
                return true;
            }
            catch (final Exception e) {
                LogHelper.wtf(e);
            }
            return false;
        }

    }

    public static class Method {

        @Nullable
        public static List<String> methods(@NonNull final Object object) {
            try {
                final List<String> methods = new ArrayList<>();
                for (final java.lang.reflect.Method method : object.getClass().getDeclaredMethods()) {
                    methods.add(method.getName());
                }
                return methods;
            }
            catch (final Exception e) {
                LogHelper.wtf(e);
            }
            return null;
        }

        @Nullable
        public static Object method(@NonNull final Object object, @NonNull final String declaredMethod, final Class[] signature) {
            try {
                java.lang.reflect.Method method = object.getClass().getDeclaredMethod(declaredMethod, signature);
                method.setAccessible(true);
                return method.invoke(null, (Object) new Class[] {});
            }
            catch (final Exception e) {
                LogHelper.wtf(e);
            }
            return null;
        }

        @Nullable
        public static Object method(@NonNull final Object object, @NonNull final String declaredMethod, final Class[] signature, final Object ... params) {
            try {
                java.lang.reflect.Method method = object.getClass().getDeclaredMethod(declaredMethod, signature);
                method.setAccessible(true);
                return method.invoke(object, params);
            }
            catch (final Exception e) {
                LogHelper.wtf(e);
            }
            return null;
        }

    }

}

File: library/src/main/java/me/shkschneider/skeleton/java/ReflectHelper.kt
package me.shkschneider.skeleton.java

import me.shkschneider.skeleton.helperx.Logger

@Deprecated("Reflection allows programmatic access to information about the methods of loaded classes.")
object ReflectHelper {

    object Field {

        fun fields(obj: Any): List<String>? {
            return fields(obj::class.java)
        }

        fun fields(cls: Class<Any>): List<String>? {
            try {
                return cls.declaredFields.map { it.name }
            } catch (e: Exception) {
                Logger.wtf(e)
            }
            return null
        }

        fun field(obj: Any, declaredField: String): Any? {
            return field(obj::class.java, declaredField)
        }

        fun field(cls: Class<Any>, declaredField: String): Any? {
            try {
                val field = cls.getDeclaredField(declaredField)
                field.isAccessible = true
                return field.get(cls)
            } catch (e: Exception) {
                Logger.wtf(e)
            }
            return null
        }

        fun field(obj: Any, declaredField: String, value: Any): Boolean {
            return field(obj::class.java, declaredField, value)
        }

        fun field(cls: Class<Any>, declaredField: String, value: Any): Boolean {
            try {
                val field = cls.getDeclaredField(declaredField)
                field.isAccessible = true
                field.set(cls, value)
                return true
            } catch (e: Exception) {
                Logger.wtf(e)
            }
            return false
        }

    }

    object Method {

        private val EMPTY_SIGNATURE = arrayOf<Class<*>>()

        fun methods(obj: Any): List<String>? {
            return methods(obj::class.java)
        }

        fun method(obj: Any, declaredMethod: String, signature: Array<Class<*>>): Any? {
            return method(obj::class.java, declaredMethod, signature, EMPTY_SIGNATURE)
        }

        fun method(obj: Any, declaredMethod: String, signature: Array<Class<*>>, vararg params: Any): Any? {
            return method(obj::class.java, declaredMethod, signature, *params)
        }

        fun methods(cls: Class<*>): List<String>? {
            try {
                return cls.declaredMethods.map { it.name }
            } catch (e: Exception) {
                Logger.wtf(e)
            }
            return null
        }

        fun method(cls: Class<*>, declaredMethod: String, signature: Array<Class<*>>, vararg params: Any = EMPTY_SIGNATURE): Any? {
            try {
                val method = cls.getDeclaredMethod(declaredMethod, *signature)
                method.isAccessible = true
                return method.invoke(cls, *params)
            } catch (e: Exception) {
                Logger.wtf(e)
            }
            return null
        }

    }

}
